javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll

```diff
@@ -233,6 +210,61 @@ module TaintTracking {
     }
   }
 
+  /**
+   * A taint propagating data flow edge caused by the builtin array functions.
+   */
+  private class ArrayFunctionTaintStep extends AdditionalTaintStep {
+    DataFlow::CallNode call;
+
+    ArrayFunctionTaintStep() {
+      this = call
+    }
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      // `array.map(function (elt, i, ary) { ... })`: if `array` is tainted, then so are
+      // `elt` and `ary`; similar for `forEach`
+      exists (string name, Function f, int i |
+        (name = "map" or name = "forEach") and
+        (i = 0 or i = 2) and
+        call.getArgument(0).analyze().getAValue().(AbstractFunction).getFunction() = f and
+        pred.(DataFlow::SourceNode).getAMethodCall(name) = call and
+        succ = DataFlow::parameterNode(f.getParameter(i))
+      )
+      or
+      // `array.map` with tainted return value in callback
+      exists (DataFlow::FunctionNode f |
+        call.(DataFlow::MethodCallNode).getMethodName() = "map" and
+        call.getArgument(0) = f and // Require the argument to be a closure to avoid spurious call/return flow
+        pred = f.getAReturn() and
+        succ = call
+      )
+      or
+      // `array.push(e)`, `array.unshift(e)`: if `e` is tainted, then so is `array`.
+      exists (string name |
+        name = "push" or
+        name = "unshift" |
+        pred = call.getAnArgument() and
+        succ.(DataFlow::SourceNode).getAMethodCall(name) = call
+      )
+      or
+      // `e = array.pop()`, `e = array.shift()`, or similar: if `array` is tainted, then so is `e`.
+      exists (string name |
+        name = "pop" or
+        name = "shift" or
+        name = "slice" or
+        name = "splice" |
+        pred.(DataFlow::SourceNode).getAMethodCall(name) = call and
```

Comment:
We don't normally backtrack the predecessor to its local source; only the successor needs to follow backward data flow in case of object mutation. Ordinary data flow will carry the taint from the source to the predecessor node anyway.

It might interfere with how flow through refinement nodes and guard sanitizers work (as it might bypass them), and create slight inconsistencies in path explanations. I'd prefer that we just do
`pred = call.(DataFlow::MethodCall).getReceiver()`.

Reply:
Good catch, amended as `call.(DataFlow::MethodCallNode).calls(pred, name)`